### PR TITLE
Allow AuthComponent and friends to work with user object implementing…

### PR DESCRIPTION
**Pull request by ADmad**
… ArrayAccess.

Refs #8240

### CODE REVIEW

File: src/Auth/BaseAuthorize.php [hunk 1]
```diff
@@ -57,7 +57,7 @@ public function __construct(ComponentRegistry $registry, array $config = [])
     /**
      * Checks user authorization.
      *
-     * @param array $user Active user data
+     * @param array\ArrayAccess $user Active user data
```
Comment (antograssiot) [inline src/Auth/BaseAuthorize.php]: missing `|` ?

**Comment by lorenzo** [comment]
Thanks!

File: src/Auth/Storage/StorageInterface.php [hunk 2]
```diff
@@ -30,10 +30,10 @@ public function read();
     /**
      * Write user record.
      *
-     * @param array $user User record.
+     * @param array|\ArrayAccess $user User record.
      * @return void
      */
-    public function write(array $user);
+    public function write($user);
```
Comment (markstory) [inline src/Auth/Storage/StorageInterface.php]: This could cause issues for userland extensions. They will be required to update their code now.

Comment (ADmad) [inline src/Auth/Storage/StorageInterface.php]: I did [comment](https://github.com/cakephp/cakephp/issues/8240#issuecomment-186300201) about the typehint change (albeit for AuthComponent::setUser()) on the original issue.

Comment (lorenzo) [inline src/Auth/Storage/StorageInterface.php]: I'm starting to dislike typehitnts :)

Comment (markstory) [inline src/Auth/Storage/StorageInterface.php]: If we're going to stick to the BC guidelines this should probably be reverted.

Comment (lorenzo) [inline src/Auth/Storage/StorageInterface.php]: So, the only solution would be to call a different method a deprecate this one?

Comment (lorenzo) [inline src/Auth/Storage/StorageInterface.php]: This would be a solid precedent for avoiding typehints :P

Comment (markstory) [inline src/Auth/Storage/StorageInterface.php]: Or add a new interface that doesn't have the typehint.

Comment (ADmad) [inline src/Auth/Storage/StorageInterface.php]: Adding a new interface won't help if someone has for example extended `SessionStorage` and overridden its `write()` method.

@lorenzo I would like to share your dislike :)

**Comment by chinpei215** [comment]
@ADmad Not the same as @markstory's PR, this would not a backward comptible change :(
If we release this change, users cannot make a plugin that works on both 3.2 and 3.3. Overriding the method as `write($user)`, the plugin no longer works 3.2 or earlier without E_STRICT errors. Overriding the method as `write(array $user)`, the plugin no longer works 3.3 or later without E_STRICT errors.
@markstory's PR has compatibility actually. They can declare the method as `parse($url, $method = '')`. It would work on the both versions.

**Comment by lorenzo** [comment]
I think we can just tell people upgrading they can silence E_STRICT while plugins or your own extensions are being updgraded

**Comment by ADmad** [comment]
@chinpei215 Both cases would throw an E_STRICT error so imo it's the same.

**Comment by chinpei215** [comment]
@markstory's PR causes a temporary breaking. After the plugin developer solve the declaration, the plugin works fine on both 3.2 and 3.3. This PR causes a permanent breaking.

**Comment by ADmad** [comment]
Okay I see your point, but I don't see why a particular plugin version needs to support all 3.x versions. Plugin dev can bump plugin version after making it 3.3 compatible.

**Comment by chinpei215** [comment]
Hm. So we can change typehint in the minor release? This typehint change itself is not so dangerous as long as we don't pass ArrayObject in the core.

**Comment by ADmad** [comment]
@chinpei215 The real problem is while mismatch in type hinting in a child class overriding parent class method causes E_STRICT error, mismatch in type hinting in method when implementing an interface it causes fatal error :(

**Comment by chinpei215** [comment]
@ADmad Indeed.

**Comment by ADmad** [comment]
So what is everyone's preference, use a new method instead of `write()` or use a new interface instead of `SessionInterface`?

The cons of using a new method is the name won't match with corresponding `read()`.

**Comment by markstory** [comment]
Can we continue not supporting ArrayObject?

**Comment by ADmad** [comment]
Based on the comments on original issue I thought the consensus was to support ArrayObject. But I am fine if this PR was reverted.

**Comment by lorenzo** [comment]
I think it is ok to leave it without the typehint

**Comment by chinpei215** [comment]
I googled `"Cake" "implements StorageInterface"`. Only 10 results hit and there are no such plugins actually. So I am OK this time. But I think we need to make guidelines about type hinting.

**Pull request merged**